File: 1570-final-prices-with-a-special-discount-in-a-shop/final-prices-with-a-special-discount-in-a-shop.cpp
class Solution {
public:
    vector<int> finalPrices(vector<int>& prices) {
        vector<int> ans(prices.size());
        int n=prices.size();
        stack<int> st;
        for(int i=n-1;i>=0;i--){
            while(!st.empty() && st.top()>prices[i])st.pop();

            ans[i]=st.empty()?prices[i]:prices[i]-st.top();

            st.push(prices[i]);
        }
        return ans;
    }
};

File: 1570-final-prices-with-a-special-discount-in-a-shop/final-prices-with-a-special-discount-in-a-shop.java
class Solution {
    public int[] finalPrices(int[] prices) {
        Integer n=prices.length;
        int [] ans=new int[n];
         Stack<Integer> st =new Stack();
        for(int i=n-1;i>=0;i--){
            while(!st.isEmpty() && st.peek()>prices[i])st.pop();

            ans[i]=st.isEmpty()?prices[i]:prices[i]-st.peek();

            st.push(prices[i]);
        }
        return ans;
    }
}